### ISSUE TITLE
Nondeterministic Starting Inventory

### ISSUE THREAD

**Issue by skaldarnar** [Terasology/LightAndShadow#117]
Sometimes the player starts with a nearly empty inventory with only the LaS Staff, and sometimes the player is spawned with the starting inventory of Combat System (various weapons).

**Comment by ahv15** [comment]
I want to work on this issue so what would be optimal here spawning only with the LaS Staff or spawning with the starting inventory of Combat System.

**Comment by skaldarnar** [comment]
Hm, I think what exactly should be in the starting inventory isn't the biggest problem here. I think for the fun of it the combat weapons are nice to have (the staff itself is a bit .. boring). Eventually, we want to replace those weapons with LaS-specific ones, but this is a different task.

There is a [StartingInventoryComponent](https://github.com/Terasology/Inventory/blob/develop/src/main/java/org/terasology/logic/inventory/StartingInventoryComponent.java) in https://github.com/Terasology/Inventory - I'm not sure whether this is used in LaS. 

In any case, the main issue here is that the inventory is non-deterministic, and this might hint at some more substantial problem in the foundational modules... Although a workaround to hard-override the inventory in this game mode may work, I'd appreciate some investigation where this random behavior comes from.

**Comment by ahv15** [comment]
Ok sure, this is actually part of a GSoC project so I was just hoping to go through some issues assosciated with it and solve some of them if possible before the actually project (if i get selected).

**Comment by ahv15** [comment]
@skaldarnar So I looked into it and what I understood was that player.prefab from the CombatSystem module initializes the StartingInventoryComponent to have the various weapons that are possible and the StartingInventorySystem in the Inventory module adds the weapons that had been initialized from the player.prefab to the inventory of the player when the player spawns. Furthermore the LASSystem in the LightAndShadow module removes all the items in the players inventory on player spawn and the TeleporterSystem in the LightAndShadow module gives the single magic staff to the player when teleporting.  As the event handlers in these systems don't have priority and are activated at different times when different events are passed to the player I assume that due to lack of priority there might not be any ordering as to how this happens, maybe sometimes one event is passed first and sometimes another event is passed first so one inventory might overide another inventory which leads to the random ordering of the inventory. I still have to take a closer look as to when which event is being consumed by the players but I think this should be the rough reason for the random inventory.

**Comment by ahv15** [comment]
As this issue is a part of a GSoC project I think it would be best if I left this issue to be fixed in the project, as a more concrete StartingInventory is needed with various other resources to build defenses and as I have isolated the problem (hopefully) it would be easier to decide the exact resources that we would want the inventory to consist of during the project and solidify that as our StartingInventory.

**Comment by skaldarnar** [comment]
Please go ahead if you want - the proposal phase did not even start, and the students can (and should) adapt if some parts of the project are already done.

The theory of events being handled with same priority in non-deterministic order sounds reasonable. I think it should be simple enough to lower the priority of the starting inventory handler in Inventory. Not sure about the interaction of all the other events, though. Nice work figuring that out already!

**Comment by ahv15** [comment]
I am refraining from solving this issue right now as the exact starting inventory still has to be decided as part of the project and once that is done I could probably decide to override the StartingInventory given by the player.prefab so that we could have a customized inventory for LaS which could help in building defensive buildings.
